engine/runtime-compiler/src/main/java/org/enso/compiler/context/NameResolutionAlgorithm.java

```diff
@@ -0,0 +1,61 @@
+package org.enso.compiler.context;
+
+import java.util.Optional;
+import org.enso.compiler.MetadataInteropHelpers;
+import org.enso.compiler.core.ConstantsNames;
+import org.enso.compiler.core.ir.Name;
+import org.enso.compiler.data.BindingsMap;
+import org.enso.compiler.pass.analyse.AliasAnalysis$;
+import org.enso.compiler.pass.analyse.alias.Info;
+import org.enso.compiler.pass.resolve.GlobalNames$;
+import scala.Option;
+
+/**
+ * Encapsulates the logic for resolving a variable name.
+ *
+ * <p>The same logic is needed in two places: 1. in the runtime ({@link
+ * org.enso.interpreter.runtime.IrToTruffle.processName}), 2. in the type checker ({@link
+ * org.enso.compiler.pass.analyse.types.TypeInference.processName}).
+ *
+ * <p>To ensure that all usages stay consistent, they should all rely on the logic implemented in
+ * this class, customizing it to the specific needs of the context in which it is used.
+ *
+ * @param <ResultType> The result type of the resolution process.
+ * @param <LocalNameLinkType> The type describing a link to a local name in a current scope.
+ *     Depending on the context this may be a Graph.Link (in the compiler) or a FramePointer (in the
+ *     runtime).
+ */
+public abstract class NameResolutionAlgorithm<ResultType, LocalNameLinkType> {
+  public final ResultType resolveName(Name.Literal name) {
+    Info.Occurrence occurrenceMetadata =
+        MetadataInteropHelpers.getMetadata(name, AliasAnalysis$.MODULE$, Info.Occurrence.class);
+    var maybeLocalLink = findLocalLink(occurrenceMetadata);
+    if (maybeLocalLink.isDefined()) {
+      return resolveLocalName(maybeLocalLink.get());
+    }
+
+    Optional<BindingsMap.Resolution> global =
+        MetadataInteropHelpers.getOptionalMetadata(
+            name, GlobalNames$.MODULE$, BindingsMap.Resolution.class);
+    if (global.isPresent()) {
```

Comment:
Explained above already.

IMO the `Optional` makes it much clearer. But if you are really against it, I can remove it - but then I'd keep `getRequiredMetadata` and `getOptionalMetadata` where one would do the nullcheck itself. 

I really don't like littering all the code with `null` checks in places where I normally expect the value to be present (but then this may be violated in some weird scenarios and the resulting NPE is annoying to debug).

Reply:
You don't need (all those devil's `null` checks). Just write `Optional.of(value)` and you can continue the _fancy FP_ style of `.map.filter.map`.